packages/core/src/indexing-store/historical.ts

```diff
@@ -1,363 +1,893 @@
+    encoding,
+    schema,
+    namespaceInfo,
+    db,
+  });
+
+  common.logger.debug({
+    service: "indexing",
+    msg: `Using a ${Math.round(maxSizeBytes / (1024 * 1024))} mB cache.`,
+  });
+
+  /** True if the cache contains the complete state of the store. */
+  let isCacheFull = _isCacheFull;
+
+  /** Number of rows in cache. */
+  let cacheSize = 0;
+  /** Estimated number of bytes used by cache. */
+  let cacheSizeBytes = 0;
+  /** LRU counter. */
+  let totalCacheOps = 0;
+
+  for (const tableName of Object.keys(tables)) {
+    storeCache[tableName] = {
+      insert: {},
+      update: {},
+      find: {},
+    };
+  }
+
+  const getCacheId = (id: UserId, tableName: string): string | number => {
+    if (tables[tableName].table.id[" scalar"] === "hex")
+      return padHex(id as Hex, {
+        size: Math.ceil(((id as Hex).length - 2) / 2),
+        dir: "left",
+      }).toLowerCase();
+    if (typeof id === "bigint") return `#Bigint.${id}`;
+    return id;
+  };
+
+  /**
+   * Validates and transforms a record to be equivalent to a record
+   * written and read from the db.
+   */
+  const sanitizeRecord = (record: UserRecord, tableName: string) => {
+    for (const [columnName, column] of Object.entries(
+      tables[tableName].table,
+    )) {
+      // optional columns are null
+      if (isOptionalColumn(column) && record[columnName] === undefined) {
+        record[columnName] = null;
+      }
+      // hex is lowercase byte encoded
+      if (
+        (isScalarColumn(column) || isReferenceColumn(column)) &&
+        column[" scalar"] === "hex" &&
+        typeof record[columnName] === "string"
+      ) {
+        record[columnName] = padHex(record[columnName] as Hex, {
+          size: Math.ceil(((record[columnName] as Hex).length - 2) / 2),
+          dir: "left",
+        }).toLowerCase();
+      }
+    }
 
-      return decodeRow(row, table, kind);
+    validateRecord({
+      record,
+      table: tables[tableName].table,
+      schema,
+      skipValidate: false,
     });
-  },
-  createMany: async ({
-    tableName,
-    data,
-  }: {
-    tableName: string;
-    data: UserRecord[];
-  }) => {
-    const table = (schema[tableName] as { table: Table }).table;
-
-    const rows: DatabaseRecord[] = [];
-
-    for (let i = 0, len = data.length; i < len; i += MAX_BATCH_SIZE) {
-      await db.wrap({ method: `${tableName}.createMany` }, async () => {
-        const createRows = data
-          .slice(i, i + MAX_BATCH_SIZE)
-          .map((d) => encodeRow(d, table, kind));
-
-        const _rows = await db
-          .withSchema(namespaceInfo.userNamespace)
-          .insertInto(tableName)
-          .values(createRows)
-          .returningAll()
-          .execute()
-          .catch((err) => {
-            throw parseStoreError(err, data.length > 0 ? data[0] : {});
-          });
+  };
+
+  const shouldFlush = () => cacheSizeBytes > maxSizeBytes;
+
+  const flush = createQueue<void, boolean | void>({
```

Comment:
What's the purpose of the queue?

Reply:
To prevent race conditions from user-land concurrency